tfprotov5/tf5server/server.go

```diff
@@ -763,6 +813,58 @@ func (s *server) UpgradeResourceState(ctx context.Context, protoReq *tfplugin5.U
 	return protoResp, nil
 }
 
+func (s *server) UpgradeResourceIdentity(ctx context.Context, protoReq *tfplugin5.UpgradeResourceIdentity_Request) (*tfplugin5.UpgradeResourceIdentity_Response, error) {
+	rpc := "UpgradeResourceIdentity"
+	ctx = s.loggingContext(ctx)
+	ctx = logging.RpcContext(ctx, rpc)
+	ctx = logging.ResourceContext(ctx, protoReq.TypeName)
+	ctx = s.stoppableContext(ctx)
+	logging.ProtocolTrace(ctx, "Received request")
+	defer logging.ProtocolTrace(ctx, "Served request")
+
+	req := fromproto.UpgradeResourceIdentityRequest(protoReq)
+
+	ctx = tf5serverlogging.DownstreamRequest(ctx)
+
+	// TODO: Remove this check and error in preference of
+	// s.downstream.UpgradeResourceIdentity below once ProviderServer interface
+	// implements this RPC method.
+	// nolint:staticcheck
+	resourceIdentityProviderServer, ok := s.downstream.(tfprotov5.ProviderServerWithResourceIdentity)
+	if !ok {
+		logging.ProtocolError(ctx, "ProviderServer does not implement UpgradeResourceIdentity")
+
+		protoResp := &tfplugin5.UpgradeResourceIdentity_Response{
+			Diagnostics: []*tfplugin5.Diagnostic{
+				{
+					Severity: tfplugin5.Diagnostic_ERROR,
+					Summary:  "Provider UpgradeResourceIdentity Not Implemented",
+					Detail: "A UpgradeResourceIdentity call was received by the provider, however the provider does not implement the call. " +
+						"Either upgrade the provider to a version that implements resource identity support or this is a bug in Terraform that should be reported to the Terraform maintainers.",
+				},
+			},
+		}
+
+		return protoResp, nil
+	}
+
+	// TODO: Update this to call downstream once optional interface is removed
+	// resp, err := s.downstream.UpgradeResourceIdentity(ctx, req)
+	resp, err := resourceIdentityProviderServer.UpgradeResourceIdentity(ctx, req)
+
+	if err != nil {
+		logging.ProtocolError(ctx, "Error from downstream", map[string]interface{}{logging.KeyError: err})
+		return nil, err
+	}
+
+	tf5serverlogging.DownstreamResponse(ctx, resp.Diagnostics)
+	logging.ProtocolData(ctx, s.protocolDataDir, rpc, "Response", "UpgradedResourceIdentity", resp.UpgradedIdentity)
```

Comment:
The last argument of this function needs to be a `DynamicValue` [to log properly](https://github.com/hashicorp/terraform-plugin-go/blob/71a1984af153b30eed462647fbc7241d7247bb45/internal/logging/protocol_data.go#L51-L59) (it's an `any` type 🙁 , so you'll want to pass in `resp.UpgradedIdentity.IdentityData`). Perhaps also a nil check before doing that, since anyone can implement `terraform-plugin-go` 😆